src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/activities/ActivitiesService.kt

```diff
@@ -110,6 +111,39 @@ class ActivitiesService(
     }
   }
 
+  fun deallocate(allocationEvent: AllocationDomainEvent) {
+    activitiesApiService.getAllocation(allocationEvent.allocationId).let { allocation ->
+      mappingService.getMappingGivenActivityScheduleId(allocationEvent.scheduleId).let { mapping ->
+
+        val telemetryMap = mutableMapOf(
+          "allocationId" to allocation.id.toString(),
+          "offenderNo" to allocation.prisonerNumber,
+          "bookingId" to allocation.bookingId.toString(),
+        )
+
+        val nomisResponse = try {
+          nomisApiService.deallocate(
+            mapping.nomisCourseActivityId,
+            allocation.bookingId!!,
+            EndOffenderProgramProfileRequest(
+              endDate = allocation.endDate!!,
+              endReason = allocation.deallocatedReason, // TODO probably will need a mapping
+              // endComment = allocation.?, // TODO could put something useful in here
+            )
+          )
+        } catch (e: Exception) {
+          telemetryClient.trackEvent("activity-deallocate-failed", telemetryMap)
+          throw e
+        }
+
+        val mapWithNomisId = telemetryMap
+          .plus(Pair("offenderProgramReferenceId", nomisResponse.offenderProgramReferenceId.toString()))
```

Comment:
Why re-assign to a new value given the map is mutable?

Reply:
good point, have changed to just put to the existing map